src/lfortran/semantics/ast_symboltable_visitor.cpp

```diff
@@ -374,7 +374,132 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
         return r;
     }
 
+    void populate_implicit_dictionary(const AST::Function_t &x, std::map<std::string, ASR::ttype_t*> &implicit_dictionary) {
+        if (compiler_options.implicit_typing) {
+            //if --implicit-typing flag is used
+
+            //then define i-n as integer and rest all as real
+
+            for (char ch='i'; ch<='n'; ch++) {
+                implicit_dictionary[std::string(1, ch)] = LFortran::ASRUtils::TYPE(ASR::make_Integer_t(al, x.base.base.loc, 4, nullptr, 0));
+            }
+
+            for (char ch='o'; ch<='z'; ch++) {
+                implicit_dictionary[std::string(1, ch)] = LFortran::ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc, 4, nullptr, 0));
+            }
+
+            for (char ch='a'; ch<='h'; ch++) {
+                implicit_dictionary[std::string(1, ch)] = LFortran::ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc, 4, nullptr, 0));
+            }
+        } else {
+            //if --implicit-typing flag is not used
+            //populate all characters as nullptr
+
+            for (char ch='a'; ch<='z'; ch++) {
+                implicit_dictionary[std::string(1, ch)] = nullptr;
+            }
+        }
+    }
+
+    void print_implicit_dictionary(std::map<std::string, ASR::ttype_t*> &implicit_dictionary) {
+        std::cout << "Implicit Dictionary: " << std::endl;
+        for (auto it: implicit_dictionary) {
+            if (it.second) {
+                std::cout << it.first << " " << ASRUtils::type_to_str(it.second) << std::endl;
+            } else {
+                std::cout << it.first << " " << "NULL" << std::endl;
+            }
+        }
+    }
+
     void visit_Function(const AST::Function_t &x) {
+        // TODO: The most common (modern) path is explicit typing, so in this
+        // path things should be as fast as possible. Currently they are slow.
+        // TODO: this needs to be moved to Common Visitor, and it needs to be
+        // made availabe when visiting the body, not the symbol table.
+        //populate the implicit_dictionary
+        populate_implicit_dictionary(x, implicit_dictionary);
```

Comment:
This should only be done if implicit typing is needed.

I think the way to go is this: if `--implicit-typing` option is used, then one can use the `implicit real ...` lines, or use the default implicit typing. If `--implicit-typing` is not used, then on can only use `implicit none` (which will be ignored) and nothing else. 


Then in the code we simply change the above to:

```suggestion
        if (compiler_options.implicit_typing) populate_implicit_dictionary(x, implicit_dictionary);
```

And in the default path (no implicit typing) there will be no overhead.